packages/langium/src/utils/caching.ts

```diff
@@ -157,10 +177,24 @@ export class DocumentCache<K, V> extends ContextCache<URI | string, K, V, string
  * If any document in the workspace changes, the whole cache is evicted.
  */
 export class WorkspaceCache<K, V> extends SimpleCache<K, V> {
-    constructor(sharedServices: LangiumSharedCoreServices) {
+
+    /**
+     * Creates a new workspace cache.
+     *
+     * @param sharedServices Service container instance to hook into document lifecycle events.
+     * @param state Optional document state on which the cache should evict.
+     * If not provided, the cache will evict on `DocumentBuilder#onUpdate`.
+     */
+    constructor(sharedServices: LangiumSharedCoreServices, state?: DocumentState) {
         super();
-        this.onDispose(sharedServices.workspace.DocumentBuilder.onUpdate(() => {
-            this.clear();
-        }));
+        if (state) {
+            this.onDispose(sharedServices.workspace.DocumentBuilder.onBuildPhase(state, () => {
+                this.clear();
+            }));
+        } else {
+            this.onDispose(sharedServices.workspace.DocumentBuilder.onUpdate(() => {
+                this.clear();
+            }));
```

Comment:
I know this form of registering disposables is used a lot in other frameworks like Theia, but IMO it's a bit unfortunate: if you don't read carefully, `this.onDispose(doSomething())` looks like `doSomething` happens only on dispose. It can be clarified by splitting it in separate statements:
```typescript
const cleanup = doSomething();
this.onDispose(cleanup);
```